titan/features/prep.py

```diff
@@ -0,0 +1,333 @@
+
+        If the agent is not hiv and time is at least the prep start time, if the agent is already on PrEP update their PrEP attributes, if the agent isn't on PrEP and is eleigible, initiate PrEP.
+
+        args:
+            model: the instance of HIVModel currently being run
+        """
+        if (
+            not self.agent.hiv
+            and model.time >= self.agent.location.params.prep.start_time
+        ):
+            if self.active:
+                self.progress(model)
+            elif self.eligible():
+                self.initiate(model)
+
+    @classmethod
+    def add_agent(cls, agent: "agent.Agent"):
+        """
+        Add an agent to the class (not instance).
+
+        Add agent to the PrEP counts by race, and add the agent to the set of new agents.
+
+        args:
+            agent: the agent to add to the class attributes
+        """
+        # set up if this is the first time being called
+        cls.counts[agent.race] += 1
+        cls.new_agents.add(agent)
+
+    @classmethod
+    def remove_agent(cls, agent):
+        """
+        Remove an agent from the class (not instance).
+
+        Decrement the prep counts by race.
+
+        args:
+            agent: the agent to remove from the class attributes
+        """
+        cls.counts[agent.race] -= 1
+
+    @classmethod
+    def update_pop(cls, model: "model.HIVModel"):
+        """
+        Update the feature for the entire population (class method).
+
+        This is called in `HIVModel.update_all_agents` before agent-level updates are made.
+
+        Resets the tracking set for `new_agents` as this is called before `update_agent`
+
+        args:
+            model: the instance of HIVModel currently being run
+        """
+        cls.new_agents = set()
+
+    def set_stats(self, stats: Dict[str, int]):
+        if self.agent in self.new_agents:
+            stats["prep_new"] += 1
+
+        if self.active:
+            stats["prep"] += 1
+            if self.type == "Inj":
+                stats["prep_injectable"] += 1
+            elif self.type == "Oral":
+                stats["prep_oral"] += 1
+
+    # =============== HELPER METHODS ===================
+
+    def initiate(self, model: "model.HIVModel", force: bool = False):
+        """
+        Place agents onto PrEP treatment. PrEP treatment assumes that the agent knows their HIV status is negative.
+
+        args:
+            model : instance of HIVModel being run
+            force : whether to force the agent to enroll instead of using the appropriate algorithm per the prep params
+        """
+        # Prep only valid for agents not on prep and are HIV negative
+        if self.active or self.agent.hiv:
+            return
+
+        params = self.agent.location.params
+
+        if force:
+            self.enroll(model.run_random)
+        else:
+            if "Racial" in params.prep.target_model:
```

Comment:
The way counts are currently set up, this only works for race, but bike rack to make this based on other categories (namely sex type?)

Reply:
I'll open an issue